In [37]:
import fitz
import pandas as pd

--------

# Medical Surgical

First: Open document with Fitz

In [13]:
doc = fitz.open(pdf_path)

In [14]:
# The output of the varaible
doc

Document('Medical Surgical.pdf')

In [15]:
# The length of the list ---- it's not a list....it's a fitz doc
len(doc)

21

In [16]:
type(doc)

fitz.fitz.Document

In [17]:
# This casts the fitz doc as a list which makes it more predictable for me.
list_of_doc = list(doc)

In [18]:
# This is extracting the text from each of the pages of the document.
# Only the relevant pages are used 
text = [page.getText('text') for page in list_of_doc[2:-3]]

text[0][:100]

'5\n4\n• 100% woven cotton\n• Individually wrapped for your convenience\n• Perforated carton for easy dis'

In [20]:
len(text)

16

In [81]:
# We can define a function to remove undesired items that can be identified by certain characters
def remove(char,lis):
    for i in lis:
        if char in i:
            print('>>>>>>>>>DROPPED ',i)
            lis.remove(i)

In [22]:
text_n=[i.replace(" \n", " ") for i in text]
#for i in text:
#    i = i.replace(" \n", " ")
#    text_n.append(i)

In [23]:
len(text_n)

16

In [24]:
text_n[:1]

['5\n4\n• 100% woven cotton\n• Individually wrapped for your convenience\n• Perforated carton for easy dispensing\n• Perfect for when only one pad is needed\n• Not made with natural rubber latex\nGauze Pads:\nItem #\nDescription\nPackaging\nSterile\n1212\n2” x 2”, 12-ply\n1/pk, 100 pk/bx, 24 bx/cs\n1312\n3” x 3”, 12-ply\n1/pk, 100 pk/bx, 24 bx/cs\n1412\n4” x 4”, 12-ply\n1/pk, 100 pk/bx, 12 bx/cs\n1212-10\n2” x 2”, 12-ply\n1/pk, 10 pk/bx, 60 bx/cs\n1313-10\n3” x 3”, 12-ply\n1/pk, 10 pk/bx, 60 bx/cs\n1414-10\n4” x 4”, 12-ply\n1/pk, 10 pk/bx, 60 bx/cs\n1212-25\n2” x 2”, 12-ply\n1/pk, 25 pk/bx, 24 bx/cs\n1313-25\n3” x 3”, 12-ply\n1/pk, 25 pk/bx, 24 bx/cs\n1414-25\n4” x 4”, 12-ply\n1/pk, 25 pk/bx, 24 bx/cs\n• 100% woven cotton\n• Lighter, economical choice for quality care\n• Not made with natural rubber latex\nBasic Care Gauze Sponges:\nItem #\nDescription\nPackaging\nSterile\n8500\n2” x 2”, 8-ply\n2/pk, 50 pk/bx, 30 bx/cs\n8501\n3” x 3”, 12-ply\n2/pk, 50 pk/bx, 24 bx/cs\n8502\n4” x 4”, 8-

In [25]:
text_=[i.split('  ') for i in text_n]
# understand the above line as
# for i in text_n:
#    _= i.split('  ')
#    text_.append(_)

In [26]:
len(text_)

16

In [27]:
text_[:1]

[['5\n4\n• 100% woven cotton\n• Individually wrapped for your convenience\n• Perforated carton for easy dispensing\n• Perfect for when only one pad is needed\n• Not made with natural rubber latex\nGauze Pads:\nItem #\nDescription\nPackaging\nSterile\n1212\n2” x 2”, 12-ply\n1/pk, 100 pk/bx, 24 bx/cs\n1312\n3” x 3”, 12-ply\n1/pk, 100 pk/bx, 24 bx/cs\n1412\n4” x 4”, 12-ply\n1/pk, 100 pk/bx, 12 bx/cs\n1212-10\n2” x 2”, 12-ply\n1/pk, 10 pk/bx, 60 bx/cs\n1313-10\n3” x 3”, 12-ply\n1/pk, 10 pk/bx, 60 bx/cs\n1414-10\n4” x 4”, 12-ply\n1/pk, 10 pk/bx, 60 bx/cs\n1212-25\n2” x 2”, 12-ply\n1/pk, 25 pk/bx, 24 bx/cs\n1313-25\n3” x 3”, 12-ply\n1/pk, 25 pk/bx, 24 bx/cs\n1414-25\n4” x 4”, 12-ply\n1/pk, 25 pk/bx, 24 bx/cs\n• 100% woven cotton\n• Lighter, economical choice for quality care\n• Not made with natural rubber latex\nBasic Care Gauze Sponges:\nItem #\nDescription\nPackaging\nSterile\n8500\n2” x 2”, 8-ply\n2/pk, 50 pk/bx, 30 bx/cs\n8501\n3” x 3”, 12-ply\n2/pk, 50 pk/bx, 24 bx/cs\n8502\n4” x 4”, 8

In [94]:
text_split = [i.split('\n') for item in text_ for i in item]
text_split[1:2]
    

[['']]

## It seems better to use the unflattened list when converting to DataFrame.

Each item on the outer list is a row and each item in the inner lists are column values.

In [95]:
df_ts = pd.DataFrame(text_split)

In [96]:
df_ts

0                                                  1    \
0                    5                                                  4   
1                                                                    None   
2                                                                    None   
3                                                                    None   
4                                                                    None   
..                 ...                                                ...   
126                                                                  None   
127   Size 14-16, Blue                                    100/bx, 2 bx/cs   
128                 33                                                 32   
129        Towelettes:  • For cleansing of face, hands, and body witho...   
130                 35                                                 34   

                                      2    \
0                     • 100% woven cotton   
1                                    None   
2                                    None   
3                                    None   
4                                    None   
..                                    ...   
126                                  None   
127                   Personal Protection   
128                                   BZK   
129  • Not made with natural rubber latex   
130                    Sting Relief Pads:   

                                                   3    \
0          • Individually wrapped for your convenience   
1                                                 None   
2                                                 None   
3                                                 None   
4                                                 None   
..                                                 ...   
126                                               None   
127   • 3-ply pleated lightweight, breathable material   
128                                               None   
129                              Cleansing Towelettes:   
130  • Pre-moistened with refreshing lemon flavored...   

                                                   4    \
0              • Perforated carton for easy dispensing   
1                                                 None   
2                                                 None   
3                                                 None   
4                                                 None   
..                                                 ...   
126                                               None   
127                                     • BFE of ≥ 99%   
128                                               None   
129                                             Item #   
130  • For post operative oral care, relief of dry ...   

                                           5    \
0    • Perfect for when only one pad is needed   
1                                         None   
2                                         None   
3                                         None   
4                                         None   
..                                         ...   
126                                       None   
127    • Meet or exceed ASTM Level 1 standards   
128                                       None   
129                                Description   
130                                  • Triples   

                                                   6    \
0                 • Not made with natural rubber latex   
1                                                 None   
2                                                 None   
3                                                 None   
4                                                 None   
..                                                 ...   
126                                               None   
127  • Available in earloop or tie on with flexible...   
128                                 

In [34]:
len(text_split)

131

In [35]:
text_split_flat = [i for item in text_split for i in item]
text_split_flat[:2]

['5', '4']

In [36]:
len(text_split_flat)

2474

In [ ]:
text_split_flat[]

In [42]:
df_tsf = pd.DataFrame(text_split_flat)

In [43]:
df_tsf

0
0                                               5
1                                               4
2                             • 100% woven cotton
3     • Individually wrapped for your convenience
4         • Perforated carton for easy dispensing
...                                           ...
2469                                     856-1000
2470                                 Medium, Bulk
2471              3/strip, 1/pk, 1500/bx, 2 bx/cs
2472                      Antiseptics & Cleansers
2473                                             

[2474 rows x 1 columns]

### Some observations
- 131 * 159 = 20829 but when we flatten the list we get 2474 items 
- There are many Null items in the unflattened list 
- If each item in unflattened list is a row then each missing item is 159 items and if they only count as one in the flattened list then that explains why the flattend list is only 2474 items

### Solution:
- elemintate all empty lists from `text_split`

In [63]:
for item in text_split:
    if item == ['']:
        text_split.remove(item)

In [65]:
text_split[1:2]

[['clinical performance',
  '• Used for a number of applications including: general wound care, dressings, prepping, wound packing, and debriding',
  '• Not made with natural rubber latex',
  'CaliberTM']]

It has been confirmed the entire list has been cleaned of empty lists.

Now, we have to remove:
- page numbers: single numerical string items at the beginning of list, sometimes the first one or two items
- descriptions: start with '•' 
- item name: ends with ':'
- category names: three simple names we will add them to the column names

Once each list in the list has been organized then we need to split them into threes which should create the three columns in a dataframe.

-------------------

In [86]:
drop_words = ['Item #',
 'Description',
 'Packaging',
 'Traditional Wound Care',
 'Traditional Wound Care',
 'clinical performance',
 'maximum performance',
 'reduction of sponges used',
 'clean placement',
 'penser box for',
 'your convenience',
 'through',
 ' maximum compression',
 'absorbency, protection, and padding to the wound site',
              ''
 ]

In [ ]:

for word in drop_words:
    remove(word,text_split_flat)
    #for i in text_split_flat:
    #    if word in i:
    #        print('>>>>>>>>>DROPPED',i)
    #        text_split_flat.remove(i)
text_split_flat
        

### The previous function drops too many items.....WE CANNOT USE IT!!!


´>>>>>>>>>DROPPED  • Polyethylene middle layer helps eliminate risk of bacterial migration and moisture strike through
´>>>>>>>>>DROPPED   maximum compression
´>>>>>>>>>DROPPED  absorbency, protection, and padding to the wound site
´>>>>>>>>>DROPPED  5
´>>>>>>>>>DROPPED  • 100% woven cotton
´>>>>>>>>>DROPPED  • Perfect for when only one pad is needed
´>>>>>>>>>DROPPED  Gauze Pads:
´>>>>>>>>>DROPPED  1212
´>>>>>>>>>DROPPED  1/pk, 100 pk/bx, 24 bx/cs
´>>>>>>>>>DROPPED  3” x 3”, 12-ply
´>>>>>>>>>DROPPED  1412

------------------

In [171]:
remove('•', text_split_flat)
#for i in text_split_flat:
#    if '•' in i:
#        print('>>>>>>>>>DROPPED',i)
#        text_split_flat.remove(i)
text_split_flat

>>>>>>>>>DROPPED  • Perfect for when only one pad is needed
>>>>>>>>>DROPPED  • Both sterile and non-sterile sponges are banded in 10’s to ensure accurate count
>>>>>>>>>DROPPED  • Virtually non-linting
>>>>>>>>>DROPPED  • Can be cut to fit any size wound without separating
>>>>>>>>>DROPPED  • Designed to wick moisture away from the site
>>>>>>>>>DROPPED  • Individually wrapped
>>>>>>>>>DROPPED  • Perforated carton for easy dispensing
>>>>>>>>>DROPPED  • Unique weave construction and finished edges prevent unraveling and linting
>>>>>>>>>DROPPED  • LF series not made with natural rubber latex
>>>>>>>>>DROPPED  • Unique weave construction and finished edges prevent unraveling and linting
>>>>>>>>>DROPPED  • Packaged in a re-sealable drawstring polybag for your convenience
>>>>>>>>>DROPPED  • Adds bulk and loft that will conform as needed, secure dressings in place, and cushion the wound
>>>>>>>>>DROPPED  • Adds bulk and loft that will conform as needed, secure dressings in place, and cu

['5',
 '4',
 'Gauze Pads:',
 'Sterile',
 '1212',
 '2” x 2”, 12-ply',
 '1/pk, 100 pk/bx, 24 bx/cs',
 '1312',
 '3” x 3”, 12-ply',
 '1/pk, 100 pk/bx, 24 bx/cs',
 '1412',
 '4” x 4”, 12-ply',
 '1/pk, 100 pk/bx, 12 bx/cs',
 '1212-10',
 '2” x 2”, 12-ply',
 '1/pk, 10 pk/bx, 60 bx/cs',
 '1313-10',
 '3” x 3”, 12-ply',
 '1/pk, 10 pk/bx, 60 bx/cs',
 '1414-10',
 '4” x 4”, 12-ply',
 '1/pk, 10 pk/bx, 60 bx/cs',
 '1212-25',
 '2” x 2”, 12-ply',
 '1/pk, 25 pk/bx, 24 bx/cs',
 '1313-25',
 '3” x 3”, 12-ply',
 '1/pk, 25 pk/bx, 24 bx/cs',
 '1414-25',
 '4” x 4”, 12-ply',
 '1/pk, 25 pk/bx, 24 bx/cs',
 'Basic Care Gauze Sponges:',
 'Sterile',
 '8500',
 '2” x 2”, 8-ply',
 '2/pk, 50 pk/bx, 30 bx/cs',
 '8501',
 '3” x 3”, 12-ply',
 '2/pk, 50 pk/bx, 24 bx/cs',
 '8502',
 '4” x 4”, 8-ply',
 '2/pk, 50 pk/bx, 12 bx/cs',
 '8503',
 '4” x 4”, 12-ply',
 '2/pk, 25 pk/bx, 24 bx/cs',
 'Non-Sterile',
 '8505',
 '2” x 2”, 8-ply',
 '200/bg, 25 bg/cs 8506',
 '2” x 2”, 12-ply',
 '200/bg, 40 bg/cs',
 '8507',
 '3” x 3”, 8-ply',
 '200/

In [186]:
prod_ind = [text_split_flat.index(i) for i in text_split_flat if ":" in i]
#for i in text_split_flat:
#    if ":" in i:
#        print(text_split_flat.index(i))

prod_ind

[2,
 31,
 77,
 116,
 170,
 185,
 195,
 217,
 230,
 262,
 287,
 301,
 309,
 321,
 333,
 341,
 360,
 365,
 387,
 398,
 413,
 444,
 450,
 458,
 485,
 509,
 534,
 624,
 678,
 686,
 715,
 720,
 734,
 752,
 757,
 772,
 786,
 798,
 810,
 815,
 830,
 844,
 863,
 872,
 886,
 930,
 947,
 965,
 967,
 993,
 1012,
 1029,
 1075,
 1097,
 1098,
 1113,
 1118,
 1125,
 1136,
 1144,
 1153,
 1185,
 1205,
 1218,
 1255,
 1256,
 1286,
 1297,
 1308,
 1320,
 1334,
 1348,
 1366,
 1367,
 1407,
 1477,
 1518,
 1550,
 1586,
 1611,
 1622,
 1639,
 1641,
 1683,
 1726,
 1728,
 1747,
 1766,
 1774,
 1778,
 1788,
 1790]